### PR TITLE
Add a Data Collector Class

### PR DESCRIPTION
**Pull request by davschneller**
Unifies the output for the receivers in the DR and the wave propagation part (since: we have a (usually sparse) gather operation of some device memory in both cases).

Plus, less stream/queue syncs for multiple receivers.

### CODE REVIEW

File: src/Parallel/DataCollector.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+#ifdef ACL_DEVICE
+#include "device.h"
+#endif
+
+namespace seissol::parallel {
+
+class DataCollector {
+  public:
+  DataCollector(const std::vector<real*>& indexDataHost,
```
Comment (Thomas-Ulrich) [inline src/Parallel/DataCollector.h:17]: don't you mean indexedDataHost?

File: src/Kernels/Receiver.cpp [hunk 2]
```diff
@@ -37,18 +37,19 @@
...
 #include "Numerical_aux/BasisFunction.h"
-
+#include "Receiver.h"
 #include <Initializer/PointMapper.h>
+#include <Monitoring/FlopCounter.hpp>
 #include <Numerical_aux/Transformation.h>
+#include <Parallel/DataCollector.h>
 #include <Parallel/MPI.h>
-#include <Monitoring/FlopCounter.hpp>
+#include <SeisSol.h>
```
Comment (Thomas-Ulrich) [inline src/Kernels/Receiver.cpp:47]: You did not modify, but it shouldn't it be
#include "SeisSol.h"
(and others)

File: src/Kernels/Receiver.cpp [hunk 3]
```diff
@@ -82,8 +83,20 @@ double seissol::kernels::ReceiverCluster::calcReceivers(  double time,
   alignas(ALIGNMENT) real timeEvaluatedAtPoint[tensor::QAtPoint::size()];
   alignas(ALIGNMENT) real timeEvaluatedDerivativesAtPoint[tensor::QDerivativeAtPoint::size()];
 #ifdef ACL_DEVICE
-  constexpr size_t dofsHostCopySize = tensor::Q::size();
-  alignas(ALIGNMENT) real dofsHostCopy[ dofsHostCopySize ];
+  if (deviceCollector.get() == nullptr) {
```
Comment (Thomas-Ulrich) [inline src/Kernels/Receiver.cpp:86]: what does it mean if there is no index in get?

File: src/DynamicRupture/Output/Builders/ReceiverBasedOutputBuilder.cpp [hunk 4]
```diff
@@ -106,12 +108,7 @@ void ReceiverBasedOutputBuilder::initBasisFunctions() {
   outputData->cellCount = elementIndices.size() + elementIndicesGhost.size();
 
 #ifdef ACL_DEVICE
-  real** preDofPtr =
-      reinterpret_cast<real**>(device::DeviceInstance::getInstance().api->allocPinnedMem(
-          sizeof(real*) * outputData->cellCount));
-  outputData->deviceDataPtr =
-      reinterpret_cast<real**>(device::DeviceInstance::getInstance().api->allocGlobMem(
-          sizeof(real*) * outputData->cellCount));
+  std::vector<real*> preDofPtr(outputData->cellCount);
```
Comment (Thomas-Ulrich) [inline src/DynamicRupture/Output/Builders/ReceiverBasedOutputBuilder.cpp:111]: any idea what means pre in preDofPtr?

File: src/Kernels/Receiver.cpp [hunk 5]
```diff
@@ -82,8 +83,20 @@ double seissol::kernels::ReceiverCluster::calcReceivers(  double time,
...
+    deviceIndices.resize(m_receivers.size());
+    std::vector<real*> dofs;
+    std::unordered_map<real*, size_t> indexMap;
```
Comment (Thomas-Ulrich) [inline src/Kernels/Receiver.cpp:89]: I don't understand what is indexMap.
I mean:
`indexMap[currentDofs] = dofs.size();`
does not really suggest `indexMap[currentDofs]` points to an index but to a size?

File: src/Parallel/DataCollector.h [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+#include "Kernels/precision.hpp"
+
+#ifdef ACL_DEVICE
+#include "device.h"
+#endif
+
+namespace seissol::parallel {
+
+class DataCollector {
```
Comment (Thomas-Ulrich) [inline src/Parallel/DataCollector.h:15]: Proposed comment (ChatGPT)
// Class for collecting and managing data, facilitating data transfers between host and device

File: src/Parallel/DataCollector.h [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+  void gather(void* stream) {
+    if (!hostAccessible && indexCount > 0) {
+#ifdef ACL_DEVICE
+      device::DeviceInstance::getInstance().algorithms.copyScatterToUniform(
+          indexDataDevice, copiedData, elemSize, elemSize, indexCount, stream);
+#endif
+    }
+  }
+
```
Comment (Thomas-Ulrich) [inline src/Parallel/DataCollector.h:57]: Here ChatGPT would comment:
// Copies uniform data to a scattered region on the device
(is that correct?)

File: src/Parallel/DataCollector.h [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  ~DataCollector() {
+    if (!hostAccessible && indexCount > 0) {
+#ifdef ACL_DEVICE
+      device::DeviceInstance::getInstance().api->freeMem(indexDataDevice);
+      device::DeviceInstance::getInstance().api->freePinnedMem(copiedData);
+#endif
+    }
+  }
+
+  void gather(void* stream) {
```
Comment (Thomas-Ulrich) [inline src/Parallel/DataCollector.h:49]: // Copies scattered data to a uniform region on the device 
  ?

File: src/Parallel/DataCollector.h [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+class DataCollector {
+  public:
+  DataCollector(const std::vector<real*>& indexDataHost,
+                size_t elemSize,
+                bool hostAccessible = false)
+      : indexCount(indexDataHost.size()), indexDataHost(indexDataHost), elemSize(elemSize),
+        hostAccessible(hostAccessible) {
+#ifndef ACL_DEVICE
+    hostAccessible = true;
```
Comment (Thomas-Ulrich) [inline src/Parallel/DataCollector.h:23]: should never happen, right?

**Review by Thomas-Ulrich — COMMENTED** [review]
uninformed review, hopefully can make the code clearer.

File: src/Kernels/Receiver.cpp (hunk 5 above)
Comment (davschneller) [reply src/Kernels/Receiver.cpp:89]: We add an element directly after that—i.e. `currentDofs` is then located at `indexMap[currentDofs]`.

File: src/Parallel/DataCollector.h (hunk 9 above)
Comment (davschneller) [reply src/Parallel/DataCollector.h:23]: If we were to use this class in a non-GPU environment, then yes

File: src/Kernels/Receiver.cpp [hunk 10]
```diff
@@ -216,3 +195,26 @@ double seissol::kernels::ReceiverCluster::calcReceivers(  double time,
...
+  // collect all data pointers to transfer. If we have multiple receivers on the same cell, we make sure to only transfer the related data once (hence, we use the `indexMap` here)
+  deviceIndices.resize(m_receivers.size());
+  std::vector<real*> dofs;
+  std::unordered_map<real*, size_t> indexMap;
+  for (size_t i = 0; i < m_receivers.size(); ++i) {
+    real* currentDofs = m_receivers[i].data.dofs();
+    if (indexMap.find(currentDofs) == indexMap.end()) {
+      // point to the current array end
+      indexMap[currentDofs] = dofs.size();
+      dofs.push_back(currentDofs);
```
Comment (Thomas-Ulrich) [inline src/Kernels/Receiver.cpp:209]: I see. Wouldn't it more readable with 
```
dofs.push_back(currentDofs);
indexMap[currentDofs] = dofs.back();
```
?

Comment (davschneller) [reply src/Kernels/Receiver.cpp:209]: Unfortunately, `dofs.back()` is then equal to `currentDofs`—we really want `indexMap[currentDofs]` to point to the index inside the `dofs` vector

File: src/Parallel/DataCollector.h (hunk 1 above)
Comment (davschneller) [reply src/Parallel/DataCollector.h:18]: It should mean "index data, acessible on the host."

**Review by Thomas-Ulrich — APPROVED** [review]

**Pull request merged**